core/lib/Thelia/Controller/Admin/ProductController.php

```diff
@@ -1804,4 +1807,65 @@ protected function arrayHasEntries(array $data, array $entries)
 
         return $status;
     }
+
+    /*
+     * PRODUCT CLONING PROCESS
+     */
+
+    /**
+     * @return mixed|\Symfony\Component\HttpFoundation\Response
+     * @throws \Exception
+     */
+    public function cloneAction()
+    {
+        if (null !== $response = $this->checkAuth($this->resourceCode, $this->getModuleCode(), [AccessManager::CREATE, AccessManager::UPDATE, AccessManager::DELETE])) {
+            return $response;
+        }
+
+        // Initialize vars
+        $cloneProductForm = $this->createForm(AdminForm::PRODUCT_CLONE);
+        $lang = $this->getSession()->getLang()->getLocale();
+
+        try {
+            // Check the form against constraints violations
+            $form = $this->validateForm($cloneProductForm, "POST");
+
+            $originalProduct = ProductQuery::create()
+                ->findPk($form->getData()['productId']);
+
+            // Build and dispatch product clone event
+
+            $productCloneEvent = new ProductCloneEvent(
+                $form->getData()['newRef'],
+                $lang,
+                $originalProduct
+            );
+            $this->dispatch(TheliaEvents::PRODUCT_CLONE, $productCloneEvent);
+
+            // Build and dispatch file clone event
+
+            $fileCloneEvent = new FileCloneEvent(
+                $originalProduct->getId(),
+                $productCloneEvent->getClonedProduct()
+            );
+            $this->dispatch(TheliaEvents::FILE_CLONE, $fileCloneEvent);
```

Comment:
File cloning event should not be triggered from the controller, but from the action.